### ISSUE TITLE
Dynamic mask failed - triggered about 500 times for the same object

### ISSUE THREAD

**Issue by Yonny24** [VorlonCD/bi-aidetection#221]
Hi,

Please can you advise me on the dynamic mask settings?
I thought I had this set up ok but my wife moves these plants each night and always detects this one as a cat. (sprinklers were going mad last night!!).

Probably just the wrong settings. So I need it to be ignored after a couple of times and possibly reset after 24 hours. 
Do I need to change 'Cleat History'?

![image](https://user-images.githubusercontent.com/64428686/112981237-46ab7980-9131-11eb-9ab3-474a39bced72.png)

Thanks!!

**Comment by VorlonCD** [comment]
you could try increasing "clear history" to 10 or 20.   Or "Create mask after" to 1.   Or right-click the mask it found and "make static" so it is always ignored.  of course this doesnt help for objects that are always moving around.

Or create an image mask of the whole area where the flower pots wonder, and set PERSON relevant object to "**Ignore image mask**" so it would always trigger.   This feature is only available in the last few builds that you don't seem to have installed yet:
[AITool-VorlonCD-v2.0.1032.zip](https://github.com/VorlonCD/bi-aidetection/files/6229020/AITool-VorlonCD-v2.0.1032.zip)

**Comment by Yonny24** [comment]
Great thanks @VorlonCD 
I'll install the latest today and note your tips.
I've increased the confidence level also.

**Comment by Yonny24** [comment]
Just installed this latest version and can't get the AI tool GUI to show up, pops up for a second, minimises to the tray and then disappears altogether. Tried rebooting, still the same.
How can I solve this? Do I need to rollback?

Thanks

**Comment by Yonny24** [comment]
> 
> 
> @Yonny24 did you unblock it first?

yep

**Comment by Yonny24** [comment]
Maybe try again

**Comment by balucanb** [comment]
Well I tried. 😁

**Comment by Yonny24** [comment]
I've increased a lot of the values here.
Only remove mask after 60 mins if object not visible in last 20 detections.
Not sure if this is the right setting but makes sense

![image](https://user-images.githubusercontent.com/64428686/113008862-ce06e600-914d-11eb-8037-3305da6dc4e2.png)

**Comment by VorlonCD** [comment]
The problem with setting some of those settings high is that the dynamic mask will stick around for a long time, potentially causing something new to be ignored when it shouldn't be.  So kind of a security risk if you care.   Maybe I wouldn't set "in the last xx detections" so high.  Set back down closer to what it was.

For the new version try a windows update, reboot and test new version again.   Then try disabling antivirus before starting AITOOL.

This is correct update procedure:

1) Re-download zip
2) Right-click > Unblock the ZIP file (**before extracting!!!**)
3) Backup existing AITOOL folder, and delete all the files and FOLDERS from the main folder except **_SETTINGS** subfolder
4) Copy the new files in place (make sure to copy the "**RUNTIMES**" subfolder from ZIP also)

@balucanb - 1032 worked for you?

**Comment by Yonny24** [comment]
@VorlonCD I'll have a go a bit later  thanks

**Comment by balucanb** [comment]
Working for me, I am having issues with the tolerances, might be a me problem vs. AI problem. I probably am not getting how they interact with each other. Been trying to get it correct for a few days no luck. Don't want to hijack your thread but here is my issue. Relevant objects (defaults) are all 30%. I only use telegram and under that I have car set to 60%. Prediction Tolerances are all at default settings. However, alerts that are less than 60% are still pushed to Telegram? My understanding is that should not be happening.

**Comment by balucanb** [comment]
> @VorlonCD I'll have a go a bit later thanks
> 
> Changed it to 15 mins for 3 detections.
> 
> Only security risk I have in this spot is the neighbours cat getting away with doing poo in my garden. But should be picked up on the grass by the AI without the masks and/or by the motion sensors.

BB gun.😂

**Comment by VorlonCD** [comment]
@balucanb - I think you still have it right.  Could be a bug still I guess.   On log tab, enable TRACE logging, then go over to history tab > right click a previous detection that was sent to telegram when it should not be and pick "Test detection again".

Look in the log for "RelevantObjectManager" or "IsRelevant" references and it should tell you why an item was not skipped. 

Also: https://hackaday.com/2019/07/12/keep-pesky-cats-at-bay-with-a-machine-learning-turret-gun/

**Comment by Tinbum1** [comment]
@balucanb **However, alerts that are less than 60% are still pushed to Telegram?**

Did you actually get it in telegram? I just tried in mine and it said in the log sent to telegram but I didnt receive it.

**Comment by balucanb** [comment]
@Tinbum1 yes sir. With a 43% match annotation. Telegram set to 60%. Ref. the log I tried what @VorlonCD said. Changed it to trace went back clicked on the one sent to telegram, tested it, went to the log filtered on that image but saw none of the entries?? My first thought is always that I jacked something up so I am going to clear all the pics out of my aifolder, clear everything out only enable that 1 camera and wait for the next one and try again...

**Comment by balucanb** [comment]
Here is the pic and I copied the log entries for that pic.
![RSyardsd 20210330_133000282](https://user-images.githubusercontent.com/34525580/113040380-54c3bf00-915e-11eb-9374-0175f160b0a0.jpg)
[log entries.txt](https://github.com/VorlonCD/bi-aidetection/files/6231494/log.entries.txt)

**Comment by Yonny24** [comment]
@VorlonCD 
Unfortunately tried the new version several times, rebooting, turning off firewall etc.
The GUI just does not work on my windows 10 dell.

Do you have a slightly earlier version but newer than mine with some extra features?

**Comment by VorlonCD** [comment]
@balucanb I'll have to set up telegram to test, maybe later today or tomorrow. 
@Yonny24 - Did you try forcing a Windows Update, and disable antivirus?   If that doesnt help, post the MOST RECENT memory dump file that should have been created here:
C:\users\Yourname\AppData\Local\CrashDumps\AITOOL.EXE.xxxx.dmp
And the most recent log from AITOOL\_Setttings\Logs.   Make sure the log doesnt have any passwords, telegram API, etc listed you care about.

**Comment by balucanb** [comment]
@VorlonCD small update. I changed all the defaults to 60% and Telegram to 60% which seems to fix but that is not how it should be working of course. I also disabled dynamic masking but as soon as a car parked in front of the house it was going off nonstop so I had to re enable. DM is set at at defaults 30%

**Comment by VorlonCD** [comment]
@Yonny24 - does this fix the crash?  (@balucanb no changes for telegram issue yet)
[AITool-VorlonCD-v2.0.1036.zip](https://github.com/VorlonCD/bi-aidetection/files/6252956/AITool-VorlonCD-v2.0.1036.zip)

If not, I'll still take the DMP and LOG files as mentioned above.

Also, just start uninstalling things you dont need on that machine.   For example, I find apps that try to help with multi montiors, manage windows or build in dell/hp bloatware often cause stability issues with apps.

And have you tried just renaming your _SETTINGS folder and starting fresh?  That would rule out something in settings that is misconfigured.   You could always revert back.   If it IS settings related, I could take zipped copy and test here.

**Comment by balucanb** [comment]
Just loaded it up, I will let you know....Thanks

**Comment by balucanb** [comment]
@VorlonCD I think this has fixed the issue, I have not had anything sent to Telegram that shouldn't have been sent, and D-masking seems to be doing it's thing as planned. DISCLAIMER- I have shut all my cameras down except the 1 and been changing settings both in AI-Tool and BI so I could focus better on issues. As is always the case with me- the "problem" may have been self-imposed, but all seems to be good now. Thanks again Batman!

**Comment by VorlonCD** [comment]
Ok, great.  Interesting, I didnt think I fixed anything to do with your issue, only Yonny's.   But maybe they were related.

**Comment by balucanb** [comment]
Still having (different ) issues..Sorry..Dropped the issue? Bug? into the new discussions thread. Want to make sure it is a bug vs. something I am doing / not doing. Thanks

**Comment by Yonny24** [comment]
@VorlonCD 

Sorry been a hectic week and will be until next week.
I'll try again with the new version shortly.

On another topic how to I prevent these water marks from the sprinkler from setting off the alert? Looks like a huge cat with a tail :)

![image](https://user-images.githubusercontent.com/64428686/114109616-fd36f900-98ab-11eb-9b0e-96f7c4edb31c.png)

**Comment by Yonny24** [comment]
Yeahh AI tool back with
AITool-VorlonCD-v2.0.1036.zip

Awesome thanks as always

**Comment by Yonny24** [comment]
What is the Default Cam for?

**Comment by VorlonCD** [comment]
Maybe go to History > Prediction Details > right-click over the "cat" and "Create static mask".   Then maybe in DYNAMIC MASK SETTINGS you could try lowering the "Remove mask after xx days if static mask has not been used" so that mask will go away in a few days  

With the latest version, you can also specific a minimum and max percentage of the image for "cat" if you want.   It may or may not help.   You can see the current "percentofimage" listed in History > Prediction details for an object.

Latest installer can always be found here now:
https://github.com/VorlonCD/bi-aidetection/tree/master/src/UI/Installer

**Comment by Yonny24** [comment]
I am wondering if I create the static mask (I do for many items like plants etc), if an object moves within that static mask is it all ignored?

This example I gave was a large area of the shot. If a cat 5 times smaller walks within that mask just made will it still capture it?

Thanks

**Comment by Yonny24** [comment]
Also a bit off topic but just some advice needed.
During heavy rain last night I was getting lots of alerts from large rain drops.
I changed the trigger min duration (make time) from 0.8s to 1.5 sec which helped but still got a few alerts.

Do you have any suggestions on how to handle these events? Do not really want to make the 'make' time any higher as it could miss fast-moving relevant objects.

![image](https://user-images.githubusercontent.com/64428686/114171981-6f8af600-990b-11eb-9d8f-68c5cf2e42e6.png)

**Comment by VorlonCD** [comment]
CAT5:  No probably wouldnt match.   In Dynamic masking settings there is a "Percent Match" number you can set which defaults to 85.  That is calculated by first getting the "part" of the new object rectangle that is inside the mask (which could be all of it).  If it IS partially or entirely inside, then it calculates the difference in size based on width and height of both the mask and the new object.

For rain, there may not be a perfect solution for that.    What if you just decrease Deepstack level to Medium or Low?   I get a lot more false positives with HIGH mode.

What if there was a feature that tried to detect an big increase in the number of images being processed and automatically paused for a while?  Like snow/rain/windy or a parade being held in your driveway.  Need someone to refine this thought for me a bit....    

`If more than XX images being processed a minute for XX minutes then PAUSE for XX minutes?`  Or `PAUSE until less than XX images per minute`

Would that logic work?

**Comment by VorlonCD** [comment]
@Yonny24 do you have a comment on this old issue before I close it without mercy?  :)  My finger is on the CLOSE button...  My itchy close button finger...  Doc says not much he can do.
